arch/risc-v/src/common/riscv_addrenv.c

```diff
@@ -91,6 +91,22 @@
 static_assert((ARCH_ADDRENV_VBASE & RV_MMU_SECTION_ALIGN) == 0,
               "Addrenv start address is not aligned to section boundary");
 
+/* Inject the Extra MMU Flags for User Text and Data (Svpbmt / T-Head MMU) */
+
+#ifndef CONFIG_ARCH_MMU_UTEXT_EXTRAFLAGS
+#  define CONFIG_ARCH_MMU_UTEXT_EXTRAFLAGS (0)
+#endif
+
+#ifndef CONFIG_ARCH_MMU_UDATA_EXTRAFLAGS
+#  define CONFIG_ARCH_MMU_UDATA_EXTRAFLAGS (0)
+#endif
+
+#define UTEXT_EXTRAFLAGS ((uint64_t)CONFIG_ARCH_MMU_UTEXT_EXTRAFLAGS << 32)
```

Comment:
Actually the entire 64-bit value is equivalent to a [RISC-V Sv39 Page Table Entry](https://five-embeddev.com/riscv-priv-isa-manual/Priv-v1.12/supervisor.html#sec:sv39), which contains:

![image](https://github.com/user-attachments/assets/3a1b564e-76b9-4bb3-9f44-fda2a6f01198)

- Bits 0 to 7: Whether the MMU Page is Read  / Write / Executable etc
- Bits 10 to 53: Physical Page Number
- Bits 53 to 63: Reserved / Svpbmt Flags / T-Head Flags

So I don't think it's safe to allow the Lower 32-bits to be tampered via the NuttX Config? Also NuttX already defines Bits 0 to 7 in the `#define` constants MMU_UTEXT_FLAGS and MMU_UDATA_FLAGS.

Then the the Upper Bits get complicated: We need to support (non-standard) T-Head MMU Flags and (standard) RISC-V Svpbmt MMU Flags:
- T-Head MMU Flags are in Bits 59 to 63 (upper 5 bits)
- Svpbmt MMU Flags are in Bits 61 and 62 (upper 3 bits)

Since T-Head and Svpbmt disagree on the MMU Bits (and we may have more MMU Bits in future), that's why I decided to configure only the Upper 32 bits.

Reply:
@pkarashchenko Just to clarify, are we suggesting to expose each T-Head and Svpbmt Flag in the NuttX Config? So we might have 7 flags like this: [arch/risc-v/src/k230/k230_mm_init.c](https://github.com/apache/nuttx/blob/master/arch/risc-v/src/k230/k230_mm_init.c#L42-L76)

```c
#ifdef CONFIG_K230_PBMT_THEAD
/* T-Head Memory Type PTE definitions */
#  define _PAGE_SEC   (1UL << 59)   /* Security */
#  define _PAGE_SHARE (1UL << 60)   /* Shareable */
#  define _PAGE_BUF   (1UL << 61)   /* Bufferable */
#  define _PAGE_CACHE (1UL << 62)   /* Cacheable */
#  define _PAGE_SO    (1UL << 63)   /* Strong Order */

#else
/* Svpbmt Memory Type PTE definitions */
#  define _PAGE_NC    (1UL << 61)
#  define _PAGE_IO    (1UL << 62)
#endif
```

Reply:
Oops I meant 14 flags in total: 7 flags for User Data, 7 flags for User Text

Reply:
Alternatively, we could group the flags similar to [Linux Kernel](https://github.com/torvalds/linux/blob/master/arch/riscv/include/asm/pgtable-64.h#L112-L140):

- NOCACHE for Non-Cacheable Memory:  Weakly-ordered, Non-cacheable, Bufferable, Shareable, Non-trustable
- PMA for Cacheable Memory:  Weakly-ordered, Cacheable, Bufferable, Shareable, Non-trustable
- IO for I/O Memory: Strongly-ordered, Non-cacheable, Non-bufferable, Shareable, Non-trustable

Then NuttX Kernel will flip on the different bits for T-Head and Svpbmt, specific to its implementation.

Reply:
Another Suggestion: For safety we could erase the Bits 33 to 53 (Reserved + Physical Page Number), so in future we could reuse the Reserved Bits for new MMU Features?

```c
#define UTEXT_EXTRAFLAGS \
  ((uint64_t)(CONFIG_ARCH_MMU_UTEXT_EXTRAFLAGS & 0xF8000000) \
  << 32)
```

Reply:
Ok, I am fine with the current approach

Reply:
I was trying to say that the cache-able vs non-cache-able region usually is a pretty standard option for MMU (not sure about buffer-able), so I was thinking of exposing option like that. I do not have strong opinion here and just more writing my thoughts. I've seen that we enabled caching (and buffering) for the kernel (code and data) unconditionally, so was thinking more why do we have an option for the user segment and what is the value of that. And expanding even more into the case when we have a loadable code and should we expose a parameter for each piece of code that we load or not and should selection be made.

Reply:
I just searched for `MMU_UDATA_FLAGS` and `MMU_UTEXT_FLAGS` in the code tree and found references in next RISC-V related files:
```
arch/risc-v/src/common/riscv_addrenv.c
arch/risc-v/src/common/riscv_exception.c
arch/risc-v/src/common/riscv_pgalloc.c
```
so current change updates only `arch/risc-v/src/common/riscv_addrenv.c` and keeps the rest unchanged.
If we are having some arch variant specific implementation maybe it would be good to create something like `arch/risc-v/src/bl808/hardware/bl808_mmu.h` and expose it via `#include "chip.h"` either to `arch/risc-v/src/common/riscv_mmu.h` modifying
```
#define MMU_UTEXT_FLAGS         (PTE_R | PTE_X | PTE_U)
#define MMU_UDATA_FLAGS         (PTE_R | PTE_W | PTE_U)

...

#define MMU_KTEXT_FLAGS         (PTE_R | PTE_X | PTE_G)
#define MMU_KDATA_FLAGS         (PTE_R | PTE_W | PTE_G)
```
to
```
#define MMU_UTEXT_FLAGS         (PTE_R | PTE_X | PTE_U | ARCH_MMU_UTEXT_FLAGS )
#define MMU_UDATA_FLAGS         (PTE_R | PTE_W | PTE_U | ARCH_MMU_UTEXT_FLAGS)

...

#define MMU_KTEXT_FLAGS         (PTE_R | PTE_X | PTE_G | ARCH_MMU_KTEXT_FLAGS)
#define MMU_KDATA_FLAGS         (PTE_R | PTE_W | PTE_G | ARCH_MMU_KDATA_FLAGS)
```
or something similar.
Most likely `MMU_UPGT_FLAGS` and `MMU_KPGT_FLAGS` may require some update as well.

On the other hand I see in `arch/risc-v/src/common/riscv_mmu.h` a block with
```
#ifdef CONFIG_ARCH_MMU_TYPE_SV39
...
#elif CONFIG_ARCH_MMU_TYPE_SV32
...
#else
```
so maybe we even do not need such a complex `arch/risc-v/src/bl808/hardware/bl808_mmu.h` inheritance schema, but can simple move all `MMU_XXXXX_FLAGS` definitions below MMU type selector and define some extra bits for those architectures combining needed set into
```
#define RV_MMU_UTEXT_FLAGS ...
#define RV_MMU_UDATA_FLAGS ...
...
```
and getting final flags defined as
```
#define MMU_UTEXT_FLAGS         (PTE_R | PTE_X | PTE_U | RV_MMU_UTEXT_FLAGS )
#define MMU_UDATA_FLAGS         (PTE_R | PTE_W | PTE_U | RV_MMU_UDATA_FLAGS)
...
```

Reply:
Having something more generic like in `k230_mm_init.c` is also fine. I mean
```
RV_MMU_TEXT_FLAGS
RV_MMU_DATA_FLAGS
RV_MMU_SHM_FLAGS
```
or
```
ARCH_MMU_TEXT_FLAGS
ARCH_MMU_DATA_FLAGS
ARCH_MMU_SHM_FLAGS
```
so we do not need separate set for user and kernel text and data